ext-src/mem-reg.patch

```diff
@@ -0,0 +1,89 @@
+diff --git a/apps/nccl/include/nccl.h b/apps/nccl/include/nccl.h
+index 7f50792..b8b146d 100644
+--- a/apps/nccl/include/nccl.h
++++ b/apps/nccl/include/nccl.h
+@@ -344,6 +344,12 @@ ncclResult_t ncclAllGather(const void* sendbuff, void* recvbuff, size_t sendcoun
+ ncclResult_t pncclAllGather(const void* sendbuff, void* recvbuff, size_t sendcount, ncclDataType_t datatype,
+                             ncclComm_t comm, cudaStream_t stream);
+ 
++/*
++ * Register/Deregister
++ */
++ncclResult_t ncclCommRegister(ncclComm_t comm, void* buff, size_t size, void** handle);
++ncclResult_t ncclCommDeregister(ncclComm_t comm, void* handle);
++ncclResult_t ncclBuffIsRegistered(ncclComm_t comm, const void* buff, size_t count, bool* registered);
+ /*
+  * Send
+  *
+diff --git a/apps/nccl/src/nccl.cu b/apps/nccl/src/nccl.cu
+index a697be2..d8497e7 100644
+--- a/apps/nccl/src/nccl.cu
++++ b/apps/nccl/src/nccl.cu
+@@ -577,6 +577,67 @@ NCCL_API ncclResult_t ncclAllGather(const void* sendbuff, void* recvbuff, size_t
+   return ncclSuccess;
+ }
+ 
++NCCL_API ncclResult_t ncclCommRegister(ncclComm_t comm, void* buff, size_t size, void** handle) {
++  size_t buffBytes = size;
++  CUdeviceptr buffBasePtr;
++  MSCCLPP_CUTHROW(cuMemGetAddressRange(&buffBasePtr, &buffBytes, (CUdeviceptr)buff));
++
++  int rank = comm->comm->bootstrap()->getRank();
++  channelKey buffKey{(void*)buffBasePtr, buffBytes};
++
++  std::vector<mscclpp::RegisteredMemory> remoteMemories;
++
++  // Creating the channels
++  auto buffIt = comm->channelScratchInfos.find(buffKey);
```

Comment:
If there's no guarantee this will be called from a single thread, we probably need to lock access to channelScratchInfos and the rest of the maps. I suggest using one lock if this function doesn't fall on the critical path.

Reply:
This map is supposed to be checked by the host process and MSCCLPP is invoked only when using one GPU per process case.